src/main/java/org/opensearch/ad/transport/ForwardADTaskTransportAction.java

```diff
@@ -26,45 +26,145 @@
 
+import static org.opensearch.ad.model.ADTask.ERROR_FIELD;
+import static org.opensearch.ad.model.ADTask.STATE_FIELD;
+import static org.opensearch.ad.model.ADTask.TASK_PROGRESS_FIELD;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.opensearch.OpenSearchStatusException;
 import org.opensearch.action.ActionListener;
 import org.opensearch.action.support.ActionFilters;
 import org.opensearch.action.support.HandledTransportAction;
+import org.opensearch.ad.model.ADTask;
 import org.opensearch.ad.model.ADTaskAction;
+import org.opensearch.ad.model.ADTaskState;
 import org.opensearch.ad.model.AnomalyDetector;
 import org.opensearch.ad.model.DetectionDateRange;
+import org.opensearch.ad.task.ADTaskCacheManager;
 import org.opensearch.ad.task.ADTaskManager;
 import org.opensearch.common.inject.Inject;
 import org.opensearch.rest.RestStatus;
 import org.opensearch.tasks.Task;
 import org.opensearch.transport.TransportService;
 
-public class ForwardADTaskTransportAction extends HandledTransportAction<ForwardADTaskRequest, AnomalyDetectorJobResponse> {
+import com.google.common.collect.ImmutableMap;
 
-    private final ADTaskManager adTaskManager;
+public class ForwardADTaskTransportAction extends HandledTransportAction<ForwardADTaskRequest, AnomalyDetectorJobResponse> {
+    private final Logger logger = LogManager.getLogger(ForwardADTaskTransportAction.class);
     private final TransportService transportService;
+    private final ADTaskManager adTaskManager;
+    private final ADTaskCacheManager adTaskCacheManager;
 
     @Inject
-    public ForwardADTaskTransportAction(ActionFilters actionFilters, TransportService transportService, ADTaskManager adTaskManager) {
+    public ForwardADTaskTransportAction(
+        ActionFilters actionFilters,
+        TransportService transportService,
+        ADTaskManager adTaskManager,
+        ADTaskCacheManager adTaskCacheManager
+    ) {
         super(ForwardADTaskAction.NAME, transportService, actionFilters, ForwardADTaskRequest::new);
         this.adTaskManager = adTaskManager;
         this.transportService = transportService;
+        this.adTaskCacheManager = adTaskCacheManager;
     }
 
     @Override
     protected void doExecute(Task task, ForwardADTaskRequest request, ActionListener<AnomalyDetectorJobResponse> listener) {
         ADTaskAction adTaskAction = request.getAdTaskAction();
         AnomalyDetector detector = request.getDetector();
         DetectionDateRange detectionDateRange = request.getDetectionDateRange();
+        String detectorId = detector.getDetectorId();
+        ADTask adTask = request.getAdTask();
 
         switch (adTaskAction) {
             case START:
+                // Start historical analysis for detector
                 adTaskManager.startHistoricalAnalysisTask(detector, detectionDateRange, request.getUser(), transportService, listener);
                 break;
-            case STOP:
+            case FINISHED:
```

Comment:
Copy from https://github.com/opensearch-project/anomaly-detection/pull/58/files#r658358045
> Would this cause issues in rolling upgrade? One set of nodes using STOP, while another uses FINISHED?

Only single entity detector use this action. Not possible to return two different actions for single entity detector.